### PR TITLE
[SPARK-47123][CORE] JDBCRDD does not correctly handle errors in getQueryOutputSchema

### PR DESCRIPTION
**Pull request by planga82**
If there is an error executing statement.executeQuery(), it's possible that another error in one of the finally statements makes us not see the main error.

```
def getQueryOutputSchema(
      query: String, options: JDBCOptions, dialect: JdbcDialect): StructType = {
    val conn: Connection = dialect.createConnectionFactory(options)(-1)
    try {
      val statement = conn.prepareStatement(query)
      try {
        statement.setQueryTimeout(options.queryTimeout)
        val rs = statement.executeQuery()
        try {
          JdbcUtils.getSchema(rs, dialect, alwaysNullable = true,
            isTimestampNTZ = options.preferTimestampNTZ)
        } finally {
          rs.close()
        }
      } finally {
        statement.close()
      }
    } finally {
      conn.close()
    }
  } 
```

### What changes were proposed in this pull request?

Avoid losing the main exception

### Why are the changes needed?

Simplify error debugging

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

In an environment where we have timeout problems executing the query.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 1]
```diff
@@ -72,13 +72,25 @@ object JDBCRDD extends Logging {
           JdbcUtils.getSchema(rs, dialect, alwaysNullable = true,
             isTimestampNTZ = options.preferTimestampNTZ)
         } finally {
-          rs.close()
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:75]: Can we use ` AutoCloseable` syntax instead? That will show the original exception as well.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:75]: +1 for the above comment.

Comment (planga82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:75]: Cool! Thank you @HyukjinKwon @dongjoon-hyun . PR updated.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**